File: app/src/main/java/com/tommasov/mg4swipenovalauncher/PreferencesManager.java
package com.tommasov.mg4swipenovalauncher;

import android.content.Context;
import android.content.SharedPreferences;

public class PreferencesManager {
    private static final String PREFS_NAME = "SwipeServicePrefs";
    private static final String KEY_PACKAGE_NAME = "packageName";
    private static final String KEY_BACK_BUTTON_VISIBILITY = "backButtonVisibility";

    private static PreferencesManager instance;
    private final SharedPreferences sharedPreferences;

    private PreferencesManager(Context context) {
        this.sharedPreferences = context.getApplicationContext()
                .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    public static synchronized PreferencesManager getInstance(Context context) {
        if (instance == null) instance = new PreferencesManager(context);
        return instance;
    }

    public void saveSelectedPackage(String packageName) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(KEY_PACKAGE_NAME, packageName);
        editor.apply();
    }

    public String getSelectedPackage() {
        return sharedPreferences.getString(KEY_PACKAGE_NAME, null);
    }

    public void saveBackButtonVisibility(String visibility) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(KEY_BACK_BUTTON_VISIBILITY, visibility);
        editor.apply();
    }

    public String getBackButtonVisibility() {
        return sharedPreferences.getString(KEY_BACK_BUTTON_VISIBILITY, "INVISIBLE");
    }

    /** Convenience: true if the floating back button is hidden */
    public boolean isBackButtonHidden() {
        return !"VISIBLE".equals(getBackButtonVisibility());
    }

    // ── Floating button position ──

    public void saveButtonPosition(int x, int y) {
        sharedPreferences.edit().putInt("btn_x", x).putInt("btn_y", y).apply();
    }

    public int getButtonX() { return sharedPreferences.getInt("btn_x", 25); }

    // ── Swipe sensitivity ──

    public int getSwipeThreshold() { return sharedPreferences.getInt("swipe_threshold", 100); }
    public int getSwipeVelocity() { return sharedPreferences.getInt("swipe_velocity", 100); }

    public void saveSwipeSensitivity(int threshold, int velocity) {
        sharedPreferences.edit().putInt("swipe_threshold", threshold).putInt("swipe_velocity", velocity).apply();
    }
    public int getButtonY() { return sharedPreferences.getInt("btn_y", 5); }
}

File: app/src/main/java/com/tommasov/mg4swipenovalauncher/PreferencesManager.kt
package com.tommasov.mg4swipenovalauncher

import android.content.Context
import android.content.SharedPreferences

/** Singleton SharedPreferences wrapper for all app settings. */
class PreferencesManager private constructor(context: Context) {

    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    var selectedPackage: String?
        get() = prefs.getString(KEY_PACKAGE_NAME, null)
        set(v) = prefs.edit().putString(KEY_PACKAGE_NAME, v).apply()

    var backButtonVisibility: String
        get() = prefs.getString(KEY_BACK_BUTTON_VISIBILITY, "INVISIBLE") ?: "INVISIBLE"
        set(v) = prefs.edit().putString(KEY_BACK_BUTTON_VISIBILITY, v).apply()

    val isBackButtonHidden: Boolean get() = backButtonVisibility != "VISIBLE"

    // ── Floating button position ──

    var buttonX: Int
        get() = prefs.getInt("btn_x", 25)
        set(v) = prefs.edit().putInt("btn_x", v).apply()

    var buttonY: Int
        get() = prefs.getInt("btn_y", 5)
        set(v) = prefs.edit().putInt("btn_y", v).apply()

    fun saveButtonPosition(x: Int, y: Int) = prefs.edit().putInt("btn_x", x).putInt("btn_y", y).apply()

    // ── Swipe sensitivity ──

    var swipeThreshold: Int
        get() = prefs.getInt("swipe_threshold", 100)
        set(v) = prefs.edit().putInt("swipe_threshold", v).apply()

    var swipeVelocity: Int
        get() = prefs.getInt("swipe_velocity", 100)
        set(v) = prefs.edit().putInt("swipe_velocity", v).apply()

    companion object {
        private const val PREFS_NAME = "SwipeServicePrefs"
        private const val KEY_PACKAGE_NAME = "packageName"
        private const val KEY_BACK_BUTTON_VISIBILITY = "backButtonVisibility"

        @Volatile private var instance: PreferencesManager? = null

        fun getInstance(context: Context): PreferencesManager =
            instance ?: synchronized(this) {
                instance ?: PreferencesManager(context).also { instance = it }
            }
    }
}
